### DISCUSSION TITLE
Binary Search Tree

### DISCUSSION THREAD

**Discussion by BLoHny** [subak-ba/cs-study#40] [자료 구조]
# BST(Binary Search Tree)

### 트리란?

![image](https://github.com/subak-ba/cs-study/assets/113517233/3d540b19-6e28-4463-a0b2-2cbaeda3e299)

트리는 항상 루트(root)에서부터 시작된다. 루트는 자식(child) 노드를 가지며, 간선(edge)으로 연결되어 있다. 자식 노드의 개수는 차수(degree)라고 하며, 크기(size)는 자신을 포함한 모든 자식 노드의 개수다. 높이(height)는 루트노드에서부터 리프(leaf)까지의 거리, 깊이(depth)는 루트에서부터 현재 노드까지의 거리다. 트리는 그 자식도 트리인 서브트리(subtree) 구성을 띤다. 레벨(level)은 0부터 시작한다. 트리는 항상 단방향(uni-directional)이다.

### 트리의 주요 용어
- 노드
  - 루트 노드 -> 트리에서의 최상위 노드
  - 부모 노드 -> 자신 노드의 위쪽 노드
  - 자식 노드 -> 자신 노드의 바로 아랫쪽 노드
  - 형제 노드 -> 같은 부모 노드를 가지는 노드
  - 리프 노드 -> 자식이 존재하지 않는 노드 : 최종 데이터
  - 경로(path): 한 노드에서 다른 한 노드에 이르는 길 사이에 있는 노드들의 순서
  - 길이(length): 출발 노드에서 도착 노드까지 거치는 간선의 개수
  - 깊이(depth): 가장 긴 루트 경로까지의 길이 ( 루트 -> 리프)
  - 레벨(level): 루트 노드(level=0)부터 노드까지 연결된 간선 수의 합
  - 높이(height): 가장 긴 루트 경로의 길이
  - 차수(degree): 각 노드의 자식의 개수
  - 트리의 차수(degree of tree): 트리의 최대 차수 
  - 크기(size): 노드의 개수
  - 너비(width): 가장 많은 노드를 갖고 있는 레벨의 크기
  - 내부 정점(internal vertex): 차수가 2 이상인 정점을 뜻한다.
  - 포레스트(forest): 서로 독립적인 트리들의 모임이다.
  - 방향 트리(directed tree): 방향을 무시하고 생각했을 때 트리인 유향 그래프는 방향 트리이다. 자료구조의 트리는 방향 트리의 일종이다

### BST의 상위 개념 이진트리란??
![image](https://github.com/subak-ba/cs-study/assets/113517233/9811e923-5481-42cc-81e7-9a45a9082b69)

차수를 2개로 제한하는 트리

이진트리의 종류
- 정 이진 트리 -> 모든 트리의 자식은 0아님 2개
- 포화 이진 트리 -> 모든 리프 노드의 높이가 같고 리프 노드가 아닌 노드는 모두 2개의 자식을 갖는다.
- 완전 이진 트리 -> 모든 리프노드의 높이가 최대 1 차이가 나고, 모든 노드의 오른쪽 자식이 있으면 왼쪽 자식이 있는 이진트리이다. 다시 말해 트리의 원소를 왼쪽에서 오른쪽으로 하나씩 빠짐없이 채워나간 형태이다. 

### 이진 트리의 순회방법
- 중위 순회
- 전위 순회
- 후외 순회
- 레벨 순서 순회
![image](https://github.com/subak-ba/cs-study/assets/113517233/133302ae-70be-4711-937e-904438791f3f)
```hi
In-order: 1 3 4 6 7 8 10 13 14
Pre-order: 8 3 1 6 4 7 10 14 13
Post-order: 1 4 7 6 3 13 14 10 8
Level-order: 8 3 10 1 6 14 4 7 13
```

## 정의
이진 트리의 일종으로 노드 기준 왼쪽 자식은 작은값, 오른쪽 자식은 큰 값을 가지는 트리이다

![image](https://github.com/subak-ba/cs-study/assets/113517233/4730cffe-a240-4e4a-9779-6bcc6ea504b4)

BST 균형 트리가 아니기에 삽입/삭제/탐색이 최고의 효율에서는 O(log N) 최악에선 O(N)이 걸리게 됩니다.
최악의 경우에 O(N)이 걸리게 되는 이유는 말그대로 균형트리가 아니기에 트리 노드들이 한쪽으로 치우치게 생성 될 수 있습니다.

### 탐색
![image](https://github.com/subak-ba/cs-study/assets/113517233/eb92cbea-29d1-4d09-af19-bf2962f3a0d4)

1. 찾고자 하는 데이터를 루트 노드와 비교합니다.
2. 크다면 왼쪽 자식노드 크다면 오른쪽 자식노드로 내려갑니다
3. 일치하는 값을 찾을때까지 2번을 재귀적으로 반복합니다
4. 리프노드에 도달시 null을 반환합니다

### 삽입
![image](https://github.com/subak-ba/cs-study/assets/113517233/338ba6a4-a6a0-4da3-bd6a-80a4db114ae4)

1. 루트 노드에서 시작하여 값이 작다면 왼쪽 크다면 오른쪽 노드로 내려갑니다
2. 1번의 크기비교를 재귀적으로 반복합니다.
3. 리프노드에 도달시 작을경우 왼쪽 클경우 리프노드의 오른쪽에 데이터를 삽입합니다.

### 삭제
- ### 삭제할 노드가 리프노드일때
![image](https://github.com/subak-ba/cs-study/assets/113517233/8f6e863a-b9be-45f4-8752-e6419558a410)

단순히 삭제만 진행하면 됨.

- ### 자식이 하나만 존재하는 노드일때
![image](https://github.com/subak-ba/cs-study/assets/113517233/01b19227-92c1-4a8d-814b-07ca321afef6)

단순히 삭제후 자식노드를 자신의 부모노드에 연결함.

- ### 왼쪽 자식, 오른쪽 자식 둘 다 있는 노드에 대한 삭제
![image](https://github.com/subak-ba/cs-study/assets/113517233/bb1da3d6-88e0-451d-bce6-5de595258d15)

** Successor 노드 : 삭제할 노드의 오른쪽 서브트리중 가장 작은 값 **

- 삭제할 노드를 찾습니다.
- 삭제할 노드의 successor 노드를 찾습니다.
- 삭제할 노드와 successor 노드의 값을 바꿉니다.
- successor 노드를 삭제합니다.

## 코드 구현
```java
public class BinarySearchTree {
    BSTNode root;

    public BinarySearchTree() {
        root = null;
    }

    void insert(int value) {
        root = insertRec(root, value);
    }

    BSTNode insertRec(BSTNode root, int value) {
        if (root == null) {
            root = new BSTNode(value);
            return root;
        }
        if (value < root.value)
            root.left = insertRec(root.left, value);
        else if (value > root.value)
            root.right = insertRec(root.right, value);

        return root;
    }

    void delete(int value) {
        root = deleteRec(root, value);
    }

    BSTNode deleteRec(BSTNode root, int value) {
        if (root == null) return root;

        if (value < root.value)
            root.left = deleteRec(root.left, value);
        else if (value > root.value)
            root.right = deleteRec(root.right, value);
        else {
            if (root.left == null)
                return root.right;
            else if (root.right == null)
                return root.left;

            root.value = minValue(root.right);
            root.right = deleteRec(root.right, root.value);
        }
        return root;
    }

    int minValue(BSTNode root) {
        int minv = root.value;
        while (root.left != null) {
            minv = root.left.value;
            root = root.left;
        }
        return minv;
    }

    void printAll() {
        jung(root);
    }

    void jung(BSTNode root) {
        if (root != null) {
            jung(root.left);
            System.out.print(root.value + " ");
            jung(root.right);
        }
    }

    public static void main(String[] args) {
        BinarySearchTree bst = new BinarySearchTree();

        // 삽입
        bst.insert(50);
        bst.insert(30);
        bst.insert(20);
        bst.insert(40);
        bst.insert(70);
        bst.insert(60);
        bst.insert(80);

        bst.printAll();
        System.out.println();

        bst.delete(20);
        bst.delete(30);
        bst.delete(50);

        bst.printAll();
    }
}


class BSTNode {
    int value;
    BSTNode left, right;

    public BSTNode(int item) {
        value = item;
        left = right = null;
    }
}
```

**Comment by onee-only** [comment]
이진트리의 삭제에서 successor를 어떻게 찾는지 설명을 보충해주실 수 있나요?

**Reply by BLoHny** [reply]
추가하였습니다.

**Comment by jangwooooo** [comment]
> 높이(height)는 현재 위치에서부터 리프(leaf)까지의 거리
> 높이(height): 가장 긴 루트 경로의 길이

사전준비에서 설명하는 용어의 개념이 다른 거 같아요

**Reply by BLoHny** [reply]
애매할 수 있다고 생각이 드네요. 반영하겠습니다.

**Comment by jangwooooo** [comment]
포화이진트리와 완전이진트리의 형태는 같나요?
정의만 봤을 때 저는 형태가 같을 거 같다고 생각돼서요.

**Reply by BLoHny** [reply]
![image](https://github.com/subak-ba/cs-study/assets/113517233/325c58bb-691a-4599-b9f9-e67b8e20dba7)
포화이진트리는 말그대로 최고 LEVEL(모든 층)이 전부 노드로 채워진 것을 의미합니다.

반대로 완전 이진트리는
![image](https://github.com/subak-ba/cs-study/assets/113517233/04f632f5-37b2-48c2-9bae-4ff6124927c7)

LEVEL - 1까지만 포화상태이고 최고 LEVEL에서 왼쪽부터 채워진 트리를 의미합니다

**Reply by jangwooooo** [reply]
그러면 포화이진트리와 완전이진트리의 차이점은 삽입 시 왼쪽 값부터 채워진다 정도인가요?

**Reply by BLoHny** [reply]
제 생각엔 포화이진과 완전이진은 이진트리에서의 상태만 나타낸 정의어이므로 가장 큰 차이는 위의 설명일 것 같습니다,

**Reply by jangwooooo** [reply]
답변 감사합니다. 🙇🏻‍♂️

**Comment by jangwooooo** [comment]
삽입 시 같은 값이 이미 존재하면 어떻게 되나요?

**Reply by BLoHny** [reply]
대체로 BST에서는 중복 value를 허용하지 않습니다
https://www.cs.usfca.edu/~galles/visualization/BST.html 이런 BST 테스터 사이트에서는 오른쪽 서브트리로 무제한 삽입을 하기는 합니다.

**Reply by jangwooooo** [reply]
아하 답변 감사힙니다.
그러면 BST뿐만 아니라 이진트리에서는 중복 값을 허용하지 않나요??

**Reply by BLoHny** [reply]
이진트리도 BST랑 비슷하게 중복 비허용을 주로 사용하고
만약 중복을 허용한다면 예시로 연결리스트와 결합하여 허용합니다.

**Reply by jangwooooo** [reply]
넵 답변 감사합니다
